drivers: can: stm32: support for dual CAN instances

martinjaeger opened the pull request:
This PR implements support for the STM32 dual CAN, where CAN1 and CAN2 share the memory for the filter banks.

Compared to the previous proposal #40531, this implementation only supports an equal split of the available 28 filter banks, which leads to the same amount of 14 filter banks per instance independent of the number of instances available in the MCU. This simplifies the driver implementation significantly and should cover most use cases.

The implementation was tested on a Nucleo F446RE and passes all tests with the following command (the board definition was updated to use `can2` instance instead of `can1` for testing):
```
scripts/twister --device-testing --device-serial /dev/ttyACM0 -p nucleo_f446re -T tests/drivers/can
```

In theory, also CAN3 available on some STM32F7 series MCUs should work if the missing devicetree nodes are added, but I don't have any hardware to test.

martinjaeger committed:
drivers: can: stm32: use DT_INST_FOREACH for driver setup

This commit unifies the driver initialization for can1 and can2
instances so that a single macro can be used.

Enabling the master clock for can2 as introduced in 8ab81b02 had to
be moved to devicetree in order to have the same CAN_STM32_CONFIG_INST
macro for all instances.

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger committed:
drivers: can: stm32: rework internal functions

Remove two functions which were only called from one place and did
not improve readibility of the code.

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger committed:
drivers: can: stm32: support for dual CAN instances

CAN1 and CAN2 share the memory for the filter banks.

This commit adds an offset for filters installed for CAN2, allowing to
use both CAN instances simultaneously.

The hardware allows to freely split the avalable 28 filters between
CAN1 and CAN2. In order to simplify the driver implementation, it only
supports an equal split of 14 filters per instance. This is the same
amount of filters as available in MCUs with only a single CAN.

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger committed:
boards: arm: nucleo_f446re: use can2 instance for testing

Choose can2 as the zephyr,canbus instance and add necessary settings
for testing of the STM32 dual CAN driver implementation.

Signed-off-by: Martin Jäger <martin@libre.solar>

str4t0m reviewed (changes requested):
Thanks for this enhancement! I have not tested it on hardware, but I'd like to highlight two points.
- one-shot mode:
 The config initialization got lost during the multi instance refactoring.
- locking of filter registers:
Locking is done per instance, but both instances access the filter registers of CAN1 (master_can).
I believe we should make sure that the filter configuration happens mutually exclusive.

towen commented:
I have tested this on hardware, and can confirm it works :)

Caveats: I am not using one-shot mode or any special filter config, but I am using heavy traffic (ISO-TP bootloader on CAN2) while simultaneously sending messages from CAN1.

martinjaeger commented:
> * one-shot mode:
>   The config initialization got lost during the multi instance refactoring.

Good spot! I must have accidentally dropped that line. It's now added back.

> * locking of filter registers:
>   Locking is done per instance, but both instances access the filter registers of CAN1 (master_can).
>   I believe we should make sure that the filter configuration happens mutually exclusive.

That's also a good point. Added a static `filter_mutex` now. This is not the most efficient way if there are 3 CAN instances (as the third one is independent), but I think that's acceptable as it's only affecting the configuration of the filters, which is usually not done very often.

henrikbrixandersen commented on drivers/can/can_stm32.c:
```diff
@@ -1163,132 +1151,81 @@ static const struct can_driver_api can_api_funcs = {
+PINCTRL_DT_INST_DEFINE(inst);                                            \
+static const struct can_stm32_config can_stm32_cfg_##inst = {            \
+	.can = (CAN_TypeDef *)DT_INST_REG_ADDR(inst),                    \
+	.master_can = (CAN_TypeDef *)DT_INST_PROP_OR(inst,               \
+		master_can_reg, DT_INST_REG_ADDR(inst)),                 \
```
Wouldn't this register _always_ be the reg base of the `can1` DTS node? Can we get rid of the separate DTS property for it now?

martinjaeger replied:
No, we should keep it, as CAN3 is an independent peripheral again and should work as well with this approach. Just couldn't test because I dont have a board with 3 CANs.

henrikbrixandersen reviewed (approved):
Looks good, one suggestion:

erwango reviewed (approved):
Good improvement, thanks!
